### PR TITLE
Add openai vlm

### PR DESCRIPTION
**Pull request by zhao0625**
- add OpenAI VLM
- include a test image and also working request code block
- add a small test

### CODE REVIEW

File: predicators/pretrained_model_interface.py [hunk 1]
```diff
@@ -249,3 +253,135 @@ def _sample_completions(
...
+    images = [PIL.Image.open("../tests/datasets/test_vlm_predicate_img.jpg")]
+
+    print("Start requesting...")
+    completions = vlm.sample_completions(prompt=prompt,
+                                         imgs=images,
+                                         temperature=0.5,
+                                         num_completions=3,
+                                         seed=0)
+    for i, completion in enumerate(completions):
+        print(f"Completion {i + 1}: \n{completion}\n")
```
Comment (NishanthJKumar) [inline predicators/pretrained_model_interface.py:387]: remove; we can show usage somewhere else/in the test

File: predicators/pretrained_model_interface.py [hunk 2]
```diff
@@ -5,16 +5,20 @@
...
+from io import BytesIO
 from typing import List, Optional
 
+import cv2
 import google
 import google.generativeai as genai
 import imagehash
 import openai
 import PIL.Image
+from tenacity import retry, stop_after_attempt, wait_random_exponential
```
Comment (NishanthJKumar) [inline predicators/pretrained_model_interface.py:21]: this needs to be added as a requirement in setup.py and potentially some other places

**Review by NishanthJKumar — COMMENTED** [review]
thanks for doing this! looks like there are a number of changes necessary before merging; I can get to them later on, or you can do them sooner if you'd like

File: predicators/pretrained_model_interface.py (hunk 1 above)
Comment (zhao0625) [reply predicators/pretrained_model_interface.py:387]: sounds good; I can move it to test? maybe a separate test function but without actually requesting?

File: predicators/pretrained_model_interface.py (hunk 2 above)
Comment (zhao0625) [reply predicators/pretrained_model_interface.py:21]: will do

**Comment by zhao0625** [comment]
sounds good! they seem quick, I will finish soon

**Comment by zhao0625** [comment]
@NishanthJKumar Thanks for finishing it up!

File: predicators/pretrained_model_interface.py [hunk 3]
```diff
@@ -159,23 +160,56 @@ def sample_completions(
...
+class OpenAIModel():
+    """Common interface with methods for all OpenAI-based models."""
+
+    def set_openai_key(self, key: Optional[str] = None) -> None:
+        """Set the OpenAI API key."""
+        if key is None:
+            assert "OPENAI_API_KEY" in os.environ
+            key = os.environ["OPENAI_API_KEY"]
+
+    @retry(wait=wait_random_exponential(min=1, max=60),
```
Comment (tomsilver) [inline predicators/pretrained_model_interface.py:172]: lol are we hackers now

**Review by tomsilver — APPROVED** [review]
LGTM!

**Pull request merged**